client/state/sites/reducer.js

```diff
@@ -138,34 +138,49 @@ export function items( state = {}, action ) {
 
 			// A site settings update may or may not include the icon property.
 			// If not, we should simply return state unchanged.
-			if ( ! settings.hasOwnProperty( 'site_icon' ) ) {
+			if ( ! settings.hasOwnProperty( 'site_icon' ) && ! settings.hasOwnProperty( 'blog_public' ) ) {
 				return state;
 			}
 
-			const mediaId = settings.site_icon;
-
-			// Similarly, return unchanged if next icon matches current value,
-			// accounting for the fact that a non-existent icon property is
-			// equivalent to setting the media icon as null
 			const site = state[ siteId ];
-			if ( ! site || ( ! site.icon && null === mediaId ) ||
-					( site.icon && site.icon.media_id === mediaId ) ) {
+			let nextSite;
+
+			if ( ! site ) {
 				return state;
 			}
 
-			let nextSite;
-			if ( null === mediaId ) {
-				// Unset icon
-				nextSite = omit( site, 'icon' );
-			} else {
-				// Update icon, intentionally removing reference to the URL,
-				// shifting burden of URL lookup to selector
+			if ( settings.hasOwnProperty( 'blog_public' ) ) {
+				// Site settings returns a numerical value for blog_public.
+				// Site state expects a true/false value for is_private.
+				const sitePrivacy = parseInt( settings.blog_public, 10 ) !== 1;
+
 				nextSite = {
 					...site,
-					icon: {
-						media_id: mediaId
-					}
+					is_private: sitePrivacy
 				};
+			} else if ( settings.hasOwnProperty( 'site_icon' ) ) {
```

Comment:
> Couldn't a settings payload have both blog_public and site_icon? In which case we'd only update one or the other because of else if. Probably better as a separate if.

I ran through some scenarios, and I don't think the settings payload will actually ever contain both a `blog_public` and `site_icon` update. When the site icon is updated, we call [`saveSiteSettings` immediately](https://github.com/Automattic/wp-calypso/blob/3ed0e7d0a658ca7656f190401fa82996885bcefe/client/my-sites/site-settings/site-icon-setting/index.jsx#L93) which automatically saves the settings with only the icon property. You can see this if you change the blog privacy (but don't save), update the icon (see the "Settings Saved" dialogue), and then try to leave the page. You'll see an error because the blog privacy setting is still queued.

Happy to update it either way just in case this behavior is changed in the future.